JAMES-2993 session refactoring

Arsnael opened the pull request:
A refactoring of the session usage, regarding the discussion in https://github.com/linagora/james-project/pull/3035#discussion_r363684700

Note: At the end, I did an attempt to refactor the `ReIndexerPerformer` to not use the `Mapper`, but didn't succeed. The problem being that the `ListeningMessageSearchIndex` is expecting `MailboxMessage` , returned by the `Mapper`, while a `MessageManager` returns a `MessageResult`. Starting to dig into this for a change sounds too time consuming and painful.

However, we agreed with @chibenwa that it was ok to let it like this for now, as `ReIndexerPerformer` is part of the `mailbox-tools-indexer` module, which makes it then a mailbox module?

But this change should allow us to not use the mapper in `MessageFastViewProjectionCorrector`, so it's still a necessary one.

chibenwa commented on mailbox/store/src/test/java/org/apache/james/mailbox/store/StoreRightManagerTest.java:
```diff
@@ -51,238 +51,372 @@
-        assertThatThrownBy(() -> storeRightManager.applyRightsCommand(mailboxPath, aclCommand, aliceSession))
-            .isInstanceOf(DifferentDomainException.class);
+    @Nested
+    class SystemRightsTest {
```
Please test `myRights` method too

chibenwa commented on mailbox/store/src/test/java/org/apache/james/mailbox/store/StoreRightManagerTest.java:
```diff
@@ -51,238 +51,372 @@
+
+            assertThat(storeRightManager.isReadWrite(aliceSession, mailbox, new Flags()))
+                .isTrue();
+        }
```
Add this test case:

```
        @Test
        void isReadWriteShouldReturnTrueWhenSystemSessionType() throws Exception {
            Mailbox mailbox = new Mailbox(MailboxPath.forUser(BOB, MailboxConstants.INBOX), UID_VALIDITY);

            assertThat(storeRightManager.isReadWrite(aliceSession, mailbox, new Flags()))
                .isTrue();
        }
```

chibenwa commented on mailbox/store/src/test/java/org/apache/james/mailbox/store/StoreRightManagerTest.java:
```diff
@@ -51,238 +51,372 @@
+        }
+
+        @Test
+        void isReadWriteShouldReturnTrueWhenSystemUserDoesNotHaveInsertOrPerformExpungeRightOnMailboxAndNullFlag() throws Exception {
```
There is no such things as a **SystemUser**  plean prefer the **WhenSystemSession** wording...

chibenwa commented on mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerTest.java:
```diff
@@ -1410,7 +1410,7 @@ void searchShouldNotReturnNoMoreDelegatedMailboxes() throws Exception {
     @Nested
```
I would like to have at least the following test:

```
USER 1 creates a mailbox and save a mail in it
Using a systemSession with a different username, I am able to read it
```

Maybe also demonstrating basic operations can be nice too (rename mbx, delete mbx, create mbx, copy msg, setFlags, delete msg, move msg, read msg, search mbx, search msg)

Arsnael replied:
Hmmm... regarding basic operations it depends. Some might be correct, but not all (I looked at createMailbox for example, it asserts that the MailboxPath belongs to the user, nothing to do with rights)

chibenwa replied:
That is where having some tests allow us to better understand the limits of this `systemSession` stuff and thus is valuable

Arsnael commented:
```
[b2b28c5b175057faf9d475e0ef6cd4bab03e307c] [ERROR] Failures: 
[b2b28c5b175057faf9d475e0ef6cd4bab03e307c] [ERROR]   UserMailboxesRoutesTest$ExceptionHandling.deleteShouldReturnOkOnMailboxDoesNotExists:876 1 expectation failed.
[b2b28c5b175057faf9d475e0ef6cd4bab03e307c] Expected status code <204> but was <500>.
[b2b28c5b175057faf9d475e0ef6cd4bab03e307c] 
[b2b28c5b175057faf9d475e0ef6cd4bab03e307c] [ERROR]   UserMailboxesRoutesTest$ExceptionHandling.deleteShouldReturnOkOnMailboxNotFoundExceptionWhenRemovingMailboxes:926 1 expectation failed.
[b2b28c5b175057faf9d475e0ef6cd4bab03e307c] Expected status code <204> but was <500>.
```

Arsnael commented:
I did a rebase on #3041 as I needed the extra tests added in `MailboxManagerTest` to be part of the refactoring as well

Arsnael commented:
```
[98254f49d30faad65c195179c15018ae40345c4d] [ERROR] Failures: 
[98254f49d30faad65c195179c15018ae40345c4d] [ERROR]   MailboxManagerTest$BasicFeaturesTests.renameMailboxByIdShouldChangeMailboxPathOfMailboxOfAnOtherUserWhenSystemSession:2254 
[98254f49d30faad65c195179c15018ae40345c4d] Expecting:
[98254f49d30faad65c195179c15018ae40345c4d]  <#private:Username{localPart=user_1, domainPart=Optional.empty}:mbx1>
[98254f49d30faad65c195179c15018ae40345c4d] to be equal to:
[98254f49d30faad65c195179c15018ae40345c4d]  <#private:Username{localPart=user_1, domainPart=Optional.empty}:mbx2>
[98254f49d30faad65c195179c15018ae40345c4d] but was not.
[98254f49d30faad65c195179c15018ae40345c4d] [ERROR]   MailboxManagerTest$BasicFeaturesTests.renameMailboxShouldChangeMailboxPathOfMailboxOfAnOtherUserWhenSystemSession:2239 
[98254f49d30faad65c195179c15018ae40345c4d] Expecting:
[98254f49d30faad65c195179c15018ae40345c4d]  <#private:Username{localPart=user_1, domainPart=Optional.empty}:mbx1>
[98254f49d30faad65c195179c15018ae40345c4d] to be equal to:
[98254f49d30faad65c195179c15018ae40345c4d]  <#private:Username{localPart=user_1, domainPart=Optional.empty}:mbx2>
[98254f49d30faad65c195179c15018ae40345c4d] but was not.
[98254f49d30faad65c195179c15018ae40345c4d] [ERROR] Errors: 
[98254f49d30faad65c195179c15018ae40345c4d] [ERROR]   MailboxManagerTest$BasicFeaturesTests.copyMessagesShouldCopyAllMessagesFromOneMailboxToAnOtherOfASameUser:2037 » IllegalState
[98254f49d30faad65c195179c15018ae40345c4d] [ERROR]   MailboxManagerTest$BasicFeaturesTests.copyMessagesShouldCopyMessagesWithMailboxesNotBelongingToSameUserWhenSystemSession:2353 » IllegalState
[98254f49d30faad65c195179c15018ae40345c4d] [ERROR]   MailboxManagerTest$BasicFeaturesTests.copyMessagesShouldCopyOnlyOneMessageFromOneMailboxToAnOtherOfASameUser:2075 » IllegalState
[98254f49d30faad65c195179c15018ae40345c4d] [ERROR]   MailboxManagerTest$BasicFeaturesTests.moveMessagesShouldMoveAllMessagesFromOneMailboxToAnOtherOfASameUser:1915 » IllegalState
[98254f49d30faad65c195179c15018ae40345c4d] [ERROR]   MailboxManagerTest$BasicFeaturesTests.moveMessagesShouldMoveMessagesWithMailboxesNotBelongingToSameUserWhenSystemSession:2317 » IllegalState
[98254f49d30faad65c195179c15018ae40345c4d] [ERROR]   MailboxManagerTest$BasicFeaturesTests.moveMessagesShouldMoveOnlyOneMessageFromOneMailboxToAnOtherOfASameUser:1949 » IllegalState
```

Arsnael commented:
```
[9ae597bdc7c775e576de8f765c314bcaacdfd135] [ERROR] Failures: 
[9ae597bdc7c775e576de8f765c314bcaacdfd135] [ERROR]   LocalDeliveryTest.mailShouldBeWellDeliveredByDefaultToUserWhenVirtualHostingIsTurnedOff:119 
[9ae597bdc7c775e576de8f765c314bcaacdfd135] Wanted but not invoked:
[9ae597bdc7c775e576de8f765c314bcaacdfd135] messageManager.appendMessage(
[9ae597bdc7c775e576de8f765c314bcaacdfd135]     <any org.apache.james.mailbox.MessageManager.AppendCommand>,
[9ae597bdc7c775e576de8f765c314bcaacdfd135]     <any org.apache.james.mailbox.MailboxSession>
[9ae597bdc7c775e576de8f765c314bcaacdfd135] );
[9ae597bdc7c775e576de8f765c314bcaacdfd135] -> at org.apache.james.transport.mailets.delivery.LocalDeliveryTest.mailShouldBeWellDeliveredByDefaultToUserWhenVirtualHostingIsTurnedOff(LocalDeliveryTest.java:119)
[9ae597bdc7c775e576de8f765c314bcaacdfd135] Actually, there were zero interactions with this mock.
[9ae597bdc7c775e576de8f765c314bcaacdfd135] 
[9ae597bdc7c775e576de8f765c314bcaacdfd135] [ERROR]   LocalDeliveryTest.mailShouldBeWellDeliveredByDefaultToUserWhenVirtualHostingIsTurnedOn:98 
[9ae597bdc7c775e576de8f765c314bcaacdfd135] Wanted but not invoked:
[9ae597bdc7c775e576de8f765c314bcaacdfd135] messageManager.appendMessage(
[9ae597bdc7c775e576de8f765c314bcaacdfd135]     <any org.apache.james.mailbox.MessageManager.AppendCommand>,
[9ae597bdc7c775e576de8f765c314bcaacdfd135]     <any org.apache.james.mailbox.MailboxSession>
[9ae597bdc7c775e576de8f765c314bcaacdfd135] );
[9ae597bdc7c775e576de8f765c314bcaacdfd135] -> at org.apache.james.transport.mailets.delivery.LocalDeliveryTest.mailShouldBeWellDeliveredByDefaultToUserWhenVirtualHostingIsTurnedOn(LocalDeliveryTest.java:98)
[9ae597bdc7c775e576de8f765c314bcaacdfd135] Actually, there were zero interactions with this mock.
[9ae597bdc7c775e576de8f765c314bcaacdfd135] 
[9ae597bdc7c775e576de8f765c314bcaacdfd135] [ERROR]   ToRecipientFolderTest.folderParameterShouldBeInboxByDefault:155 
[9ae597bdc7c775e576de8f765c314bcaacdfd135] Wanted but not invoked:
[9ae597bdc7c775e576de8f765c314bcaacdfd135] messageManager.appendMessage(
[9ae597bdc7c775e576de8f765c314bcaacdfd135]     <any org.apache.james.mailbox.MessageManager.AppendCommand>,
[9ae597bdc7c775e576de8f765c314bcaacdfd135]     <any org.apache.james.mailbox.MailboxSession>
[9ae597bdc7c775e576de8f765c314bcaacdfd135] );
[9ae597bdc7c775e576de8f765c314bcaacdfd135] -> at org.apache.james.transport.mailets.delivery.ToRecipientFolderTest.folderParameterShouldBeInboxByDefault(ToRecipientFolderTest.java:155)
[9ae597bdc7c775e576de8f765c314bcaacdfd135] Actually, there were zero interactions with this mock.
[9ae597bdc7c775e576de8f765c314bcaacdfd135] 
[9ae597bdc7c775e576de8f765c314bcaacdfd135] [ERROR]   ToRecipientFolderTest.folderParameterShouldIndicateDestinationFolder:140 
[9ae597bdc7c775e576de8f765c314bcaacdfd135] Wanted but not invoked:
[9ae597bdc7c775e576de8f765c314bcaacdfd135] messageManager.appendMessage(
[9ae597bdc7c775e576de8f765c314bcaacdfd135]     <any org.apache.james.mailbox.MessageManager.AppendCommand>,
[9ae597bdc7c775e576de8f765c314bcaacdfd135]     <any org.apache.james.mailbox.MailboxSession>
[9ae597bdc7c775e576de8f765c314bcaacdfd135] );
[9ae597bdc7c775e576de8f765c314bcaacdfd135] -> at org.apache.james.transport.mailets.delivery.ToRecipientFolderTest.folderParameterShouldIndicateDestinationFolder(ToRecipientFolderTest.java:140)
[9ae597bdc7c775e576de8f765c314bcaacdfd135] Actually, there were zero interactions with this mock.
[9ae597bdc7c775e576de8f765c314bcaacdfd135] 
[9ae597bdc7c775e576de8f765c314bcaacdfd135] [ERROR]   ToRecipientFolderTest.folderParameterShouldWorkWhenVirtualHostingIsTurnedOff:174 
[9ae597bdc7c775e576de8f765c314bcaacdfd135] Wanted but not invoked:
[9ae597bdc7c775e576de8f765c314bcaacdfd135] messageManager.appendMessage(
[9ae597bdc7c775e576de8f765c314bcaacdfd135]     <any org.apache.james.mailbox.MessageManager.AppendCommand>,
[9ae597bdc7c775e576de8f765c314bcaacdfd135]     <any org.apache.james.mailbox.MailboxSession>
[9ae597bdc7c775e576de8f765c314bcaacdfd135] );
[9ae597bdc7c775e576de8f765c314bcaacdfd135] -> at org.apache.james.transport.mailets.delivery.ToRecipientFolderTest.folderParameterShouldWorkWhenVirtualHostingIsTurnedOff(ToRecipientFolderTest.java:174)
[9ae597bdc7c775e576de8f765c314bcaacdfd135] Actually, there were zero interactions with this mock.
```

chibenwa commented on mailbox/api/src/main/java/org/apache/james/mailbox/SessionProvider.java:
```diff
@@ -42,6 +46,18 @@
      */
     MailboxSession createSystemSession(Username userName);
```
Let's not take care of this now, but I am reluctant of having a `username` associated with a system session (it's a lie, and a non-sense).

Maybe we could have a `Tenant` API (empty) with a `UserTenant` and a `ServiceTenant` thus:
 - This would replace the `Session Type` + `Username` variable combinaison
 - Right resolution could be done with a pattern matching like syntax
 - And finally it would remove a user of `Username` class for something that **is not the name of an user** 

If done, it should be in a separate PR as it solves a connex problem.

Feedback welcomed on this proposal.

Arsnael replied:
Might need to discuss a bit more about details but in the idea I agree, a system session shouldn't depend on a regular username, it's non sense indeed.

Better in an other PR yes, this one is big enough already

rouazana replied:
Sorry I don't get why it's a lie and a non-sense, can you explain me?

chibenwa replied:
Well a service do not look like much of a user to me, it's different context.

For instance, I don't want to care if my service is valid regarding `virtualHosting`, but I want to care about it for my usernames.  

Also, using a systemSession with a username makes think that you are doing impersonation, which is not enough to act on a messageId/mailboxId (as you do not have enougth information to know who you should impersonnate - root cause for this change-set). The system session is basicaly here trusted for any mailbox-api operation regardless of the 'username'.

rouazana replied:
Thanks. So you should have 2 kinds of session? A user session carrying a username and a system session without it?

chibenwa replied:
Yes either 2 sub-interface of MailboxSession, but I suspect abstracting things away behind a tenant might simplify things...

chibenwa replied:
I just gave it a second or third thought:

With have a third use case in addition from "regular" user session, system service session: we have "delegated user" use case. Currently the "delegated user" part is fully ignored as it results in a 'standard user session'. Which from an audit perspective is terrible: if Bob have enabled delegation to Alice, and that a fraudulous action happened on Bob account, we are not able to identify who caused the action!

While I believe 3 MailboxSession sub APIs would be a big burdown for such a central API, working with *composition* with the use of a `Tenant` class could greatly help us!

My 2 cents...

mbaechler replied:
Could you explain what you mean by Tenant? I'm not sure how this word matches the concept we have here.

Arsnael commented:
Squashed for easier reviews !

chibenwa commented on mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:
```diff
@@ -700,9 +700,12 @@ private boolean hasChildIn(Mailbox parentMailbox, List<Mailbox> mailboxesWithPat
 
     private Stream<MailboxId> getInMailboxes(ImmutableSet<MailboxId> inMailboxes, MailboxSession session) throws MailboxException {
-       if (inMailboxes.isEmpty()) {
+        if (inMailboxes.isEmpty()) {
```
I'd like to discuss the relevance of doing this `[]` sanitizing at the mailbox level.

This looks like a JMAP specific concern thus I would be in favor of handling this at the JMAP level.

(Other pull request?)

chibenwa replied:
In the meantime, perhaps it makes no sense to have 'empty' mailboxes when we do a search with a system session here.

Can I suggest throwing an illegalArgumentException?

chibenwa commented on mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:
```diff
@@ -411,7 +411,7 @@ private boolean userHasLookupRightsOn(Mailbox mailbox, MailboxSession session) t
 
     private void assertMailboxPathBelongToUser(MailboxSession mailboxSession, MailboxPath mailboxPath) throws MailboxException {
-        if (!mailboxPath.belongsTo(mailboxSession)) {
+        if (mailboxSession.getType() != MailboxSession.SessionType.System && !mailboxPath.belongsTo(mailboxSession)) {
```
assertMailboxPathBelongToUser is now wrong.

`assertSessionHasAdministrativeRightsOn(MailboxSession mailboxSession, MailboxPath mailboxPath)` ?

chibenwa commented on mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:
```diff
@@ -521,7 +521,7 @@ private void validateDestinationPath(MailboxPath newMailboxPath, MailboxSession
 
     private void assertIsOwner(MailboxSession mailboxSession, MailboxPath mailboxPath) throws MailboxNotFoundException {
-        if (!mailboxPath.belongsTo(mailboxSession)) {
+        if (mailboxSession.getType() != MailboxSession.SessionType.System && !mailboxPath.belongsTo(mailboxSession)) {
```
This is the same condition than for create.

Please extract it to a well named sub-method (sessionHasAdministrativeAccessOn ?)

chibenwa replied:
Idem naming `assertIsOwner` is now outdated.

chibenwa commented on mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java:
```diff
@@ -442,7 +446,7 @@ public Mailbox deleteMailbox(MailboxId mailboxId, MailboxSession session) throws
                 throw new MailboxNotFoundException(mailboxId);
             }
-            assertIsOwner(session, mailbox.generateAssociatedPath());
+            assertSessionHasAdministrativeAccessOn(session, mailbox.generateAssociatedPath());
```
Before we were throwing a "MailboxNotFoundException" in order not to leak that the mailbox exists.

Now, we are returning an InsufficientRightsException. 

This was done on purpose. I won't be surprised that some 'delegation' JMAP integration test breaks.

Can you find a way to still return the right exception here and above (by passing an exception supplier to assertSessionHasAdministrativeAccessOn ?)? Can we adda test in MbxManager for this?

Arsnael replied:
Hm went a bit fast here indeed... I guess it's not necessary to add tests for that, just modify the ones concerned in MailboxManagerTest, like instead of always check that the exception thrown is of `MailboxException` instance, directly check those two more refined ones

chibenwa replied:
Yes

Arsnael replied:
Actually took care of it regarding the tests in the other PR https://github.com/linagora/james-project/pull/3041/commits/5afbebec357b55d751dd2d45e1b2e731c734924b

Arsnael commented:
```
[ERROR]   ObjectStorageBlobStoreAWSNamespaceTest.saveConcurrentlyWithNonPreExistingBucketShouldNotFail » Execution
```
test this please

Arsnael commented:
```
[ERROR]   WebAdminUtilsTest.serverShouldBeAbletoStartConcurrently:38 » NotTerminated
```
test this please

Arsnael commented:
```
[d35482887877a5004436d03dc61081b4121cfb86] [ERROR] Failures: 
[d35482887877a5004436d03dc61081b4121cfb86] [ERROR]   MemorySendMDNMethodTest>SendMDNMethodTest.sendMDNShouldSendAMDNBackToTheOriginalMessageAuthor:319 1 expectation failed.
[d35482887877a5004436d03dc61081b4121cfb86] JSON path [0][1].list[0].headers.X-JAMES-MDN-JMAP-MESSAGE-ID doesn't match.
[d35482887877a5004436d03dc61081b4121cfb86] Expected: 1
[d35482887877a5004436d03dc61081b4121cfb86]   Actual: null
[d35482887877a5004436d03dc61081b4121cfb86] 
[d35482887877a5004436d03dc61081b4121cfb86] [ERROR]   MemorySetMailboxesMethodTest>SetMailboxesMethodTest.updateShouldFilterOwnerACL:1549 
[d35482887877a5004436d03dc61081b4121cfb86] Expecting:
[d35482887877a5004436d03dc61081b4121cfb86]   <{owner=aeiklprstwx, username@domain.tld=aeiklprstwx}>
[d35482887877a5004436d03dc61081b4121cfb86] not to contain key:
[d35482887877a5004436d03dc61081b4121cfb86]   <username@domain.tld>
```

Arsnael commented:
Regarding the error above from the CI... Was originating from `ExtractMDNOriginalJMAPMessageId` mailet, where we were having a system session in the `findMessageIdForRFC822MessageId` method with the use of a search on the `MailboxManager` with a `MultimailboxesSearchQuery` that didn't have the `inMailboxes` argument. As we changed it to an `IllegalArgumentException` regarding the system session, it was raising an issue.

Also I spotted and fixed that we were giving a system session to an admin impersonating an user... Very dangerous.

That made me think then. Should we really say : if it's internal code it's ok to use a system session? I start to think that probably not, as I just noticed above it's easy then to make a mistake and give unecessary access to someone or something that shouldn't have it. I think we should switch all the sessions back to user sessions in the code, and ONLY use system session when we really need it Here the example would be the recomputation of a preview with only the messageId. As we don't know the user of the message we need to recompute, we need the system session. Otherwise, I think we should avoid it.

Do people agree with that?

rouazana commented:
> Do people agree with that?

I guess you are right. For me there are 4 kinds of session:
1. user explicitly authenticated, then do operations: it's a user session
2. user implicitly authenticated (think of jmap, or some particuler user-related internal operations): it's an internal user session
3. admin authenticated but impersonated as user: it's an impersonation session
4. internal operations: system sessions

1. et 2. can probably be the same kind. 3. should be handled in a similar way, but the fact that's it's an admin that initiated the session could be kept for logging and auditing purpose.

chibenwa commented:
> Regarding the error above from the CI... Was originating from ExtractMDNOriginalJMAPMessageId mailet, where we were having a system session in the findMessageIdForRFC822MessageId method with the use of a search on the MailboxManager with a MultimailboxesSearchQuery that didn't have the inMailboxes argument. As we changed it to an IllegalArgumentException regarding the system session, it was raising an issue. So I changed it to an user session, and did it so for other mailets too.

I this specific case I believe we can : 
 - list all ids of Bob mailbox
 - specify that id list when searching mails.

> Also I spotted and fixed that we were giving a system session to an admin impersonating an user... Very dangerous.

This is easily auditable by `ctrl + click` on `createSystemSession`, which I encourage serious reviewers doing. I would not be too afraid of it.

> I think we should switch all the sessions back to user sessions in the code, and ONLY use system session when we really need it Here the example would be the recomputation of a preview with only the messageId. 

2 reactions on this proposition:
 - We would then lose the knowledge that it is a service taking over the action (compromising audit track)
 - Having API usage driven by some low level details like this looks like a failure to me. Either we concider we control system session usage thus don't need access control on it, either we extend right management to it (but then I fear we create an overly complex right factory).

-------------------

Let's take an example: Imagine I gained access to rabbitMQ and found a way to inject mailbox operations in some listeners for instance...

Pretending I have a user session, I could send messages on the behalf of Bob, Alice, Cedric, etc... And I would be able to perform any operation anyway. The exposure is the same than with `createSystemSession`.

The exposure being the same, I'd rather keep an intuitive API, where service don't rely on user sessions.

Arsnael commented:
Hmmm... To me honestly ideally I would think having different levels of permission for system session, depending on the uses, is better. But it sounds like quite a complex and time-consuming refactoring, that would need further deeper discussions, while I don't think we should allow such time and resources on it for now either. So need to find a compromise.

I'm ok to say that everything internal is managed by the system session. But need to be extra careful on its use. 

I will change back the changes on mailets to use the system session and will do a final swap (again) of the use of system session in the code, to be sure there is no misuse.

Does everyone agree with this course of actions or is there any concerns left?

chibenwa commented:
I agree.

Once done please squash things to ease review.

chibenwa commented:
```
[eb1f94cd3dae869782567f464cf11ac41c9bc874] [ERROR]   MemorySetMailboxesMethodTest>SetMailboxesMethodTest.updateShouldFilterOwnerACL:1549 
[eb1f94cd3dae869782567f464cf11ac41c9bc874] Expecting:
[eb1f94cd3dae869782567f464cf11ac41c9bc874]   <{owner=aeiklprstwx, username@domain.tld=aeiklprstwx}>
[eb1f94cd3dae869782567f464cf11ac41c9bc874] not to contain key:
[eb1f94cd3dae869782567f464cf11ac41c9bc874]   <username@domain.tld>
```

Arsnael commented:
Fixed and squashed !

Arsnael commented:
CI restart

test this please

Arsnael commented:
```
[eeb694b83048572af2ae80e3a8a7f8816596acab] [ERROR] Failures: 
[eeb694b83048572af2ae80e3a8a7f8816596acab] [ERROR]   RabbitMQEventDeadLettersIntegrationTest.multipleFailedEventShouldBeStoredInDeadLetter:316 1 expectation failed.
[eeb694b83048572af2ae80e3a8a7f8816596acab] JSON path . doesn't match.
[eeb694b83048572af2ae80e3a8a7f8816596acab] Expected: a collection with size <2>
[eeb694b83048572af2ae80e3a8a7f8816596acab]   Actual: [0a3fe1ef-120e-45ff-9dbd-4f9a86b082eb]
```

Arsnael commented:
test this please

Arsnael commented:
```
[eeb694b83048572af2ae80e3a8a7f8816596acab] [ERROR] Failures: 
[eeb694b83048572af2ae80e3a8a7f8816596acab] [ERROR]   FixingGhostMailboxTest.ghostMailboxBugShouldDiscardOldContent:215 1 expectation failed.
[eeb694b83048572af2ae80e3a8a7f8816596acab] JSON path [0][1].messageIds doesn't match.
[eeb694b83048572af2ae80e3a8a7f8816596acab] Expected: a collection with size <1>
[eeb694b83048572af2ae80e3a8a7f8816596acab]   Actual: []
```
test this please

mbaechler commented:
I'm trying to catch up with the discussions happening here.

They are really interesting and we must take the time to find the right design because security is an important matter.

Usually, I'm willing to use the same concepts in the code that we have in discussions instead of trying to "map" things.

I agree we probably need to refine Session concept to express the various possible use cases.

`UserSession` should just express the fact a `User` is identified and authenticated and asked for an action. 

Delegation handling should not be express using `Session` because it's a matter of Authorization, not Identification.

Impersonation is yet another situation: you are Authenticated and Identified as a user and act on behalf of somebody else. Should it be expressed as a temporary Authorization, a different Session type or something else entirely ?

Finally, regarding `SystemSession`, it should be used only for actions unrelated to a user action and probably should contain a statically defined set of rights on creation to help auditing and minimizing the security issues related to its usage.

WDYT?

Arsnael commented:
```
[eeb694b83048572af2ae80e3a8a7f8816596acab] [ERROR] Failures: 
[eeb694b83048572af2ae80e3a8a7f8816596acab] [ERROR]   FixingGhostMailboxTest.ghostMailboxBugShouldDiscardOldContent:215 1 expectation failed.
[eeb694b83048572af2ae80e3a8a7f8816596acab] JSON path [0][1].messageIds doesn't match.
[eeb694b83048572af2ae80e3a8a7f8816596acab] Expected: a collection with size <1>
[eeb694b83048572af2ae80e3a8a7f8816596acab]   Actual: []
```

Arsnael commented:
```
[eeb694b83048572af2ae80e3a8a7f8816596acab] [ERROR] Failures: 
[eeb694b83048572af2ae80e3a8a7f8816596acab] [ERROR]   FixingGhostMailboxTest.ghostMailboxBugShouldDiscardOldContent:215 1 expectation failed.
[eeb694b83048572af2ae80e3a8a7f8816596acab] JSON path [0][1].messageIds doesn't match.
[eeb694b83048572af2ae80e3a8a7f8816596acab] Expected: a collection with size <1>
[eeb694b83048572af2ae80e3a8a7f8816596acab]   Actual: []
```

Arsnael commented:
Thank you all for your feedbacks. I think the best to do at this point is to stop this PR and open a real and more transparent discussion on the mailing list regarding this matter. I will try to sum up and propose something and we can all discuss about it. Agreed?

chibenwa commented:
What's the follow up plan for this?

@Arsnael Any chance you sum up the discussions on the ML here with the work remaining to do?

(Can you do if possible a rebase / update before the sprint starts?)

Arsnael commented:
> What's the follow up plan for this?
> 
> @Arsnael Any chance you sum up the discussions on the ML here with the work remaining to do?

Based on the ML https://www.mail-archive.com/server-dev@james.apache.org/msg64119.html...

I would say two ADRs :
* one about mailbox access outside of mailbox modules
* one about the final decision about session

Then when all of that is approved / finalized, this PR needs to be reworked, and cut. I think we could do like this:
* PR to introduce the `Actor` object, that would represent the Entity using the session
* reworking this PR to do the split User session and System session, using the Actor
* PR to add the delegated session
* PR to add the impersonated session 

WDYT?

> (Can you do if possible a rebase / update before the sprint starts?)

Ok for a rebase, but there is too many things to update / rework and we need an ADR first on session refactoring to be valid right?

Arsnael commented:
Rebased.
